src/ftl/ast/ast.js

```diff
@@ -18,87 +18,100 @@ class Resource extends NodeList {
   }
 }
 
-class Section extends NodeList {
-  constructor(key, body = [], comment = null) {
-    super(body, comment);
-    this.type = 'Section';
-    this.key = key;
-  }
-}
-
 class Entry extends Node {
   constructor() {
     super();
     this.type = 'Entry';
   }
 }
 
-class Identifier extends Node {
-  constructor(name) {
+class Message extends Entry {
+  constructor(id, value = null, attributes = null, comment = null) {
     super();
-    this.type = 'Identifier';
-    this.name = name;
+    this.type = 'Entity';
+    this.id = id;
+    this.value = value;
+    this.attributes = attributes;
+    this.comment = comment;
   }
 }
 
 class Pattern extends Node {
-  constructor(source, elements, quoted = false) {
+  constructor(elements, quoted = false) {
     super();
     this.type = 'Pattern';
-    this.source = source;
     this.elements = elements;
     this.quoted = quoted;
   }
 }
 
-class Member extends Node {
-  constructor(key, value, def = false) {
+class Expression extends Node {
+  constructor() {
     super();
-    this.type = 'Member';
-    this.key = key;
+    this.type = 'Expression';
+  }
+}
+
+class StringExpression extends Expression {
+  constructor(value) {
+    super();
+    this.type = 'StringExpression';
     this.value = value;
-    this.default = def;
   }
 }
 
-class Entity extends Entry {
-  constructor(id, value = null, traits = [], comment = null) {
+class NumberExpression extends Expression {
+  constructor(value) {
     super();
-    this.type = 'Entity';
-    this.id = id;
+    this.type = 'NumberExpression';
     this.value = value;
-    this.traits = traits;
-    this.comment = comment;
   }
 }
 
-class Placeable extends Node {
-  constructor(expressions) {
+class MessageReference extends Expression {
+  constructor(id) {
     super();
-    this.type = 'Placeable';
-    this.expressions = expressions;
+    this.type = 'MessageReference';
+    this.id = id;
   }
 }
 
-class SelectExpression extends Node {
-  constructor(expression, variants = null) {
+class ExternalArgument extends Expression {
+  constructor(id) {
+    super();
+    this.type = 'ExternalArgument';
+    this.id = id;
+  }
+}
+
+class SelectExpression extends Expression {
+  constructor(exp, variants) {
     super();
     this.type = 'SelectExpression';
-    this.expression = expression;
+    this.exp = exp;
```

Comment:
Why not `expression`?

I think it makes sense to use full names in the AST parser.  I'll update the [ASDL](https://github.com/projectfluent/syntax/blob/master/fluent.asdl) too.